File: src/uart.rs
use mmio::regs::AuxReg::*;
use mmio::regs::GpioReg::*;
use mmio::regs::MmioReg::{Aux, Gpio};
use mmio::MMIO;

pub fn uart_init() {
    // Enable mini UART
    MMIO::write_reg(Aux(Enable), 1);

    // Disable transmitter and receiver during configuration
    MMIO::write_reg(Aux(MuCntl), 0);

    // Configure UART
    MMIO::write_reg(Aux(MuIer), 0);
    MMIO::write_reg(Aux(MuLcr), 3); // Set the data size to 8 bit
    MMIO::write_reg(Aux(MuMcr), 0); // No auto flow control
    MMIO::write_reg(Aux(MuBaud), 270); // Set baud rate for 115200
    MMIO::write_reg(Aux(MuIir), 6); // No FIFOs (interrupt when receiver holds at least 1 byte

    // Map UART1 to GPIO pins
    let mut reg = MMIO::read_reg(Gpio(Gpfsel1));
    reg &= !((7 << 12) | (7 << 15)); // Clear existing settings for GPIO 14, 15
    reg |= (2 << 12) | (2 << 15); // Set to alt5 for mini UART
    MMIO::write_reg(Gpio(Gpfsel1), reg);

    // Disable pull-up/down for pins 14 and 15
    MMIO::write_reg(Gpio(Gppud), 0);
    MMIO::delay(150);
    MMIO::write_reg(Gpio(GppudClk0), (1 << 14) | (1 << 15));
    MMIO::delay(150);
    MMIO::write_reg(Gpio(GppudClk0), 0);

    // Enable the transmitter and receiver
    MMIO::write_reg(Aux(MuCntl), 3);
}

pub fn send(c: u8) {
    // Wait until we can send
    while MMIO::read_reg(Aux(MuLsr)) & 0x20 == 0 {
        MMIO::delay(1);
    }

    // Write the character to the buffer
    MMIO::write_reg(Aux(MuIo), c as u32);
}

pub fn recv() -> u8 {
    // Wait until we can receive
    let mut c = recv_nb();
    while c.is_none() {
        c = recv_nb();
    }
    c.unwrap()
}

pub fn recv_nb() -> Option<u8> {
    // Check if something to read
    if MMIO::read_reg(Aux(MuLsr)) & 0x01 == 0 {
        return None;
    }
    // Read the character from the buffer
    Some((MMIO::read_reg(Aux(MuIo)) & 0xFF) as u8)
}

File: Lab2/src/uart.c
#include "../include/gpio.h"
#include "../include/uart.h"

/**
 * Set baud rate and characteristics (115200 8N1) and map to GPIO
 */
void uart_init()
{
    register unsigned int reg;

    /* initialize UART */
    *AUX_ENABLE     |= 1;       /* enable mini UART */
    *AUX_MU_CNTL     = 0;       /* Disable transmitter and receiver during configuration. */

    *AUX_MU_IER      = 0;       /* Disable interrupt */
    *AUX_MU_LCR      = 3;       /* Set the data size to 8 bit. */
    *AUX_MU_MCR      = 0;       /* Don’t need auto flow control. */
    *AUX_MU_BAUD     = 270;     /* 115200 baud */
    *AUX_MU_IIR      = 6;       /* No FIFO */
    // *AUX_MU_IIR      = 0xc6;       /* No FIFO */

    /* map UART1 to GPIO pins */
    // GPFSEL1 register controls the function of GPIO pins
    reg = *GPFSEL1;
    reg &= ~((7<<12)|(7<<15));  /* address of gpio 14, 15 */
    reg |=   (2<<12)|(2<<15);   /* set to alt5 */

    *GPFSEL1 = reg;            

    // ensures that GPIO pins are not affected by internal resistors
    *GPPUD = 0;                 /* Disable pull-up/down */
    reg=150;
    while ( reg-- )
    { 
        asm volatile("nop"); 
    }
    
    *GPPUDCLK0 = (1<<14)|(1<<15);

    //Stabilizing the Signal to ensure Proper Configuration
    reg=150; 
    while ( reg-- )
    {
        asm volatile("nop");
    }
    
    // ensures that GPIO pins are returned to a neutral state
    *GPPUDCLK0 = 0;             /* flush GPIO setup */

    *AUX_MU_CNTL = 3;           // Enable the transmitter and receiver.
    *AUX_MU_IER  = 3;           // Enable the interrupt.
}

/**
 * Send a character
 */
void uart_send(unsigned int c)
{
    /* Wait until UART transmitter is ready to accept new data. */
    do {

        asm volatile("nop");

    } while( ! ( *AUX_MU_LSR&0x20 ));

    /* write the character to the buffer */
    *AUX_MU_IO = c;

    // ensure proper line termination
    if ( c == '\n' )
    {
        do {

            asm volatile("nop");

        } while( ! ( *AUX_MU_LSR&0x20 ));

        *AUX_MU_IO = '\r';
    }
}


/**
 * Receive a character
 */
char uart_getc() {

    char r;

    /* wait until something is in the buffer */
    do{

        asm volatile("nop");

    } while ( ! ( *AUX_MU_LSR&0x01 ) );

    /* read the data*/
    r = ( char )( *AUX_MU_IO );

    /* convert carrige return to newline */
    //standardize newline characters across different systems.
    return r == '\r' ? '\n' : r;
}

/**
 * Display a string
 */
void uart_puts(const char *s)
{
    while( *s )
    {

        uart_send(*s++);

    }
}

/**
 * Display a binary value in hexadecimal
 */
void uart_hex(unsigned int d) {
    unsigned int n;
    int c;
    for(c=28;c>=0;c-=4) {
        // get highest tetrad
        n=(d>>c)&0xF;
        // 0-9 => '0'-'9', 10-15 => 'A'-'F'
        n+=n>9?0x37:0x30;
        uart_send(n);
    }
}
